File: app/src/main/java/tr/edu/iyte/quickreply/helper/ReplyItemTouchHelperCallback.java
package tr.edu.iyte.quickreply.helper;

import android.graphics.Canvas;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;

import tr.edu.iyte.quickreply.interfaces.ReplyItemTouchHelperAdapter;
import tr.edu.iyte.quickreply.interfaces.ViewHolderInteractionListener;

public class ReplyItemTouchHelperCallback extends ItemTouchHelper.Callback {
    private static final float ALPHA_FULL = 1.0f;

    private final ReplyItemTouchHelperAdapter adapter;

    public ReplyItemTouchHelperCallback(ReplyItemTouchHelperAdapter adapter) {
        this.adapter = adapter;
    }

    @Override
    public boolean isLongPressDragEnabled() {
        return false;
    }

    @Override
    public boolean isItemViewSwipeEnabled() {
        return true;
    }

    @Override
    public int getMovementFlags(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
        final int dragFlags = 0;
        final int swipeFlags = ItemTouchHelper.START | ItemTouchHelper.END;
        return makeMovementFlags(dragFlags, swipeFlags);
    }

    @Override
    public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
        return false;
    }

    @Override
    public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
        adapter.onReplyDismissed(viewHolder.getAdapterPosition());
    }

    @Override
    public void onChildDraw(Canvas c, RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, float dX, float dY, int actionState, boolean isCurrentlyActive) {
        if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE) {
            // fade out the view as it is swiped out of the parent's bounds
            final float alpha = ALPHA_FULL - Math.abs(dX) / (float) viewHolder.itemView.getWidth();
            viewHolder.itemView.setAlpha(alpha);
            viewHolder.itemView.setTranslationX(dX);
        } else {
            super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);
        }
    }

    @Override
    public void onSelectedChanged(RecyclerView.ViewHolder viewHolder, int actionState) {
        // We only want the active item to change
        if (actionState != ItemTouchHelper.ACTION_STATE_IDLE) {
            if (viewHolder instanceof ViewHolderInteractionListener) {
                // Let the view holder know that this item is being moved or dragged
                ViewHolderInteractionListener listener = (ViewHolderInteractionListener) viewHolder;
                listener.onInteractionStart();
            }
        }

        super.onSelectedChanged(viewHolder, actionState);
    }

    @Override
    public void clearView(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
        super.clearView(recyclerView, viewHolder);

        viewHolder.itemView.setAlpha(ALPHA_FULL);

        if (viewHolder instanceof ViewHolderInteractionListener) {
            // Tell the view holder it's time to restore the idle state
            ViewHolderInteractionListener listener = (ViewHolderInteractionListener) viewHolder;
            listener.onInteractionEnd();
        }
    }
}

File: app/src/main/java/tr/edu/iyte/quickreply/helper/ReplyItemTouchHelperCallback.kt
package tr.edu.iyte.quickreply.helper

import android.graphics.Canvas
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import tr.edu.iyte.quickreply.interfaces.ReplyItemTouchHelperAdapter
import tr.edu.iyte.quickreply.interfaces.ViewHolderInteractionListener

class ReplyItemTouchHelperCallback(val adapter: ReplyItemTouchHelperAdapter)
        : ItemTouchHelper.Callback() {
    private val ALPHA_FULL = 1.0f

    override fun isItemViewSwipeEnabled() = true
    override fun isLongPressDragEnabled() = false

    override fun getMovementFlags(recyclerView: RecyclerView?, viewHolder: RecyclerView.ViewHolder?)
            = ItemTouchHelper.Callback.makeMovementFlags(0, ItemTouchHelper.START or ItemTouchHelper.END)

    override fun onMove(recyclerView: RecyclerView?, viewHolder: RecyclerView.ViewHolder?, target: RecyclerView.ViewHolder?)
            = false

    override fun onSwiped(viewHolder: RecyclerView.ViewHolder?, direction: Int)
            = adapter.onReplyDismissed(viewHolder!!.adapterPosition)

    override fun onChildDraw(c: Canvas?, recyclerView: RecyclerView?, viewHolder: RecyclerView.ViewHolder?,
                             dX: Float, dY: Float, actionState: Int, isCurrentlyActive: Boolean) {
        if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE) {
            // fade out the view as it is swiped out of the parent's bounds
            val alpha = ALPHA_FULL - Math.abs(dX) / viewHolder?.itemView!!.width.toFloat()
            viewHolder.itemView.alpha = alpha
            viewHolder.itemView.translationX = dX
        } else {
            super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
        }
    }

    override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
        // We only want the active item to change
        if (actionState != ItemTouchHelper.ACTION_STATE_IDLE)
            if (viewHolder is ViewHolderInteractionListener)
                viewHolder.onInteractionStart()

        super.onSelectedChanged(viewHolder, actionState)
    }

    override fun clearView(recyclerView: RecyclerView?, viewHolder: RecyclerView.ViewHolder?) {
        super.clearView(recyclerView, viewHolder)

        viewHolder?.itemView!!.alpha = ALPHA_FULL

        if (viewHolder is ViewHolderInteractionListener)
            viewHolder.onInteractionEnd()
    }
}